mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,964 @@
+/// space.
+static bool isRectangularAffineForLoopNest(ArrayRef<AffineForOp> loopNest) {
+  for (AffineForOp forOp : loopNest) {
+    if (!forOp.hasConstantUpperBound() || !forOp.hasConstantLowerBound())
+      return false;
+  }
+  return true;
+}
+
+/// Returns true if this entire column of `matrix` is zero.
+static bool checkColumnIsZero(ArrayRef<SmallVector<int64_t, 4>> matrix,
+                              unsigned column) {
+  for (auto &row : matrix) {
+    if (row[column] != 0)
+      return false;
+  }
+  return true;
+}
+
+/// For a memref access function AX+B, it calculates both A and B and assigns
+/// these to protected members `loopAccessMatrices`(A) and `constVector`(B).
+/// The param `indVarIdMap` is used to locate position for coefficients of X
+/// in each row of matrix A.
+void LoopInterchange::getAffineAccessMatrices(
+    AffineForOp &rootForOp, SmallVector<Operation *, 8> &loadAndStoreOps,
+    DenseMap<Value, unsigned> &indVarIdMap, unsigned AffineForOpLoopNestSize) {
+
+  this->loopAccessMatrices.clear();
+  this->constVector.clear();
+
+  unsigned numOps = loadAndStoreOps.size();
+  for (unsigned i = 0; (i < numOps); ++i) {
+    Operation *srcOp = loadAndStoreOps[i];
+    MemRefAccess srcAccess(srcOp);
+
+    AffineMap map;
+    if (AffineLoadOp loadOp = dyn_cast<AffineLoadOp>(srcAccess.opInst))
+      map = loadOp.getAffineMap();
+    else if (AffineStoreOp storeOp = dyn_cast<AffineStoreOp>(srcAccess.opInst))
+      map = storeOp.getAffineMap();
+
+    SmallVector<Value, 8> operands(srcAccess.indices.begin(),
+                                   srcAccess.indices.end());
+
+    fullyComposeAffineMapAndOperands(&map, &operands);
+    map = simplifyAffineMap(map);
+    canonicalizeMapAndOperands(&map, &operands);
+
+    // Parse each map result to construct access matrices.
+    ArrayRef<AffineExpr> mapResults = map.getResults();
+
+    uint64_t noDims = map.getNumDims();
+    uint64_t noSymbols = map.getNumSymbols();
+
+    // Number of rows in an accessMatrix = Number of dimensions
+    // in the memref object.
+    // Number of Columns = (noDims + noSymbols).
+
+    SmallVector<SmallVector<int64_t, 4>, 4> accessMatrix;
+    SmallVector<int64_t, 4> constVector(mapResults.size());
+
+    for (unsigned l = 0; l < mapResults.size(); l++) {
+
+      // Parse the l-th map result (access expr for l-th dimension of this
+      // memref) to get the l-th row of this access matrix.
+      AffineExpr mapResult = mapResults[l];
+
+      // A Row of Matrix A.
+      SmallVector<int64_t, 4> Row;
+      Row.resize(
+          std::max(AffineForOpLoopNestSize, unsigned(noDims + noSymbols)));
+
+      // Check if `mapResult` is not a constant expr.
+      // If it is a constant expr like A[5], no need to walk it.
+      // Instead, push the value in constVector and
+      // leave the Row matrix to be a vector of zeroes.
+
+      if (mapResult.getKind() == AffineExprKind::Constant) {
```

Comment:
Please don't use `getKind()` - instead use `isa<AffineConstantExpr>`.

Reply:
Had tried earlier... using `isa<AffineConstantExpr>(mapResult)` where mapResult is `AffineExpr` throws error `classof is not a member of ‘mlir::AffineConstantExpr'`.

Reply:
Will replace this and next statement with dyn_cast<> and ternary operator.

Reply:
dyn_cast<> also fails with the same error. This is because dyn_cast<> has the statement ` isa<X>(Val) ? cast<X>(Val) : nullptr`. It is because of `isa<>` that it gives the same error.

Reply:
used `mapResult.isa<AffineConstantExpr>()`. Worked! Thanks.